components/brave_shields/browser/ad_block_service.cc

```diff
@@ -97,24 +56,23 @@ AdBlockService::SourceProviderObserver::~SourceProviderObserver() {
 }
 
 void AdBlockService::SourceProviderObserver::OnChanged() {
+  auto* filter_set =
+      new rust::Box<adblock::FilterSet>(adblock::new_filter_set());
+  auto on_loaded_cb =
+      base::BindOnce(&AdBlockService::SourceProviderObserver::OnFilterSetLoaded,
+                     weak_factory_.GetWeakPtr(), base::Owned(filter_set));
   if (is_filter_provider_manager_) {
     static_cast<AdBlockFiltersProviderManager*>(filters_provider_.get())
-        ->LoadDATBufferForEngine(
-            adblock_engine_->IsDefaultEngine(),
-            base::BindOnce(&AdBlockService::SourceProviderObserver::OnDATLoaded,
-                           weak_factory_.GetWeakPtr()));
+        ->LoadFilterSetForEngine(adblock_engine_->IsDefaultEngine(), filter_set,
+                                 std::move(on_loaded_cb));
   } else {
-    filters_provider_->LoadDAT(
-        base::BindOnce(&AdBlockService::SourceProviderObserver::OnDATLoaded,
-                       weak_factory_.GetWeakPtr()));
+    filters_provider_->LoadFilterSet(filter_set, std::move(on_loaded_cb));
   }
 }
 
-void AdBlockService::SourceProviderObserver::OnDATLoaded(
-    bool deserialize,
-    const DATFileDataBuffer& dat_buf) {
-  deserialize_ = deserialize;
-  dat_buf_ = std::move(dat_buf);
+void AdBlockService::SourceProviderObserver::OnFilterSetLoaded(
+    rust::Box<adblock::FilterSet>* filter_set) {
+  filter_set_ = absl::make_optional(std::move(*filter_set));
```

Comment:
This seems very odd and confusing. Taking ownership from a pointer is not the behavior you would expect. When using `base::Owned` the callback should not know or need to know that `base::Owned` was used and in this case it definitely would need to know that. I think you want `auto* filter_set` to be a unique_ptr, get a raw pointer reference using `.get()` before you create `on_loaded_cb` (and call std::move) and then pass that as `base::Unretained` to `LoadFilterSetForEngine`

Reply:
I think this would also be simpler if you used unique_ptr instead of `absl::optional` which makes things a bit confusing here and also https://github.com/brave/brave-core/pull/19946/files#r1329348390